### DISCUSSION TITLE
Render Camera Preview smoothly

### DISCUSSION THREAD

**Discussion by JaggerJo** [AvaloniaUI/Avalonia#8120] [Q&A]
### Context:
I'm building a photo booth application with Avalonia + FuncUI. 
I get a live feed of JPEG images (as one `byte array` per frame) from my camera. Note that it takes 70 - 110 ms to retrieve a preview. So we're talking ~10 FPS. 

### Problem:
Displaying the live feed is laggy.

### Things I've tried:

1. Setting the `Source` of an Image to the current preview image when I get it. This leads to super laggy results.

2. Custom control that calls `InvalidateVisual` when a new preview is received. Draws the image in the Render method. Laggy results.

3. Custom control that calls `InvalidateVisual` at a constant interval. Also laggy results. 

Strangely the preview is smooth if/while I resize the window. 

Whats the proper way of solving this?

**Comment by timunie** [comment]
Have you tried to use a `WriteableBitmap?`

**Reply by JaggerJo** [reply]
Taking a look!

**Comment by JaggerJo** [comment]
@timunie using a Writable bitmap is better, but the control is not redrawn sometimes / gets stuck. While I resize the window (to force a re-render) the preview is smooth. 


This is the code I run when I receive a new frame and decoded it to an SKBitmap:
```fsharp

// current Image element
let image: Image = image.Current

// the byte arrays I get need to be decoded to raw pixels first. I'm doing this with an SKBitmap
let preview: SKBitmap = preview

// initialize writable bitmap if null
if image.Source = null then
    image.Source <-
        new WriteableBitmap(
            PixelSize(preview.Width, preview.Height),
            Vector(96, 96),
            PixelFormat.Rgba8888,
            AlphaFormat.Opaque)

// write new data
let writableBitmap = (image.Source :?> WriteableBitmap)
let frameBuffer = writableBitmap.Lock()
Marshal.Copy(preview.Bytes, 0, frameBuffer.Address, preview.Bytes.Leng
frameBuffer.Dispose ()

// trigger render
Dispatcher.UIThread.Post (fun _ -> image.InvalidateVisual ())
// or 
image.InvalidateVisual()
```

**Comment by JaggerJo** [comment]
Another version - This also does not redraw consistently.

```fsharp
type CameraPreviewControl () as this =
    inherit Control ()
    
    let mutable link: ICamLink option = None
    let mutable writableBitmap: WriteableBitmap = null

    member this.InitLink (camLink: ICamLink) =
        link <- Some camLink
        
    override this.Render (context: DrawingContext) =
        Dispatcher.UIThread.Post(Action(this.InvalidateMeasure), DispatcherPriority.Background);
        Dispatcher.UIThread.Post(Action(this.InvalidateVisual), DispatcherPriority.Background);
        
        match link with
        | Some link -> 
            match link.CurrentPreviewBitmap.Current with
            | Some preview ->

                // initialize writable bitmap if null
                if writableBitmap = null then
                    writableBitmap <-
                        new WriteableBitmap(
                            PixelSize(preview.Width, preview.Height),
                            Vector(96, 96),
                            PixelFormat.Rgba8888,
                            AlphaFormat.Opaque)

                let frameBuffer = writableBitmap.Lock()
                Marshal.Copy(preview.Bytes, 0, frameBuffer.Address, preview.Bytes.Length)
                frameBuffer.Dispose ()

                context.DrawImage(writableBitmap, this.Bounds)
                
                printfn "Render called"
            | None ->
                ()
        | None ->
            ()
```

**Comment by JaggerJo** [comment]
bug? https://github.com/AvaloniaUI/Avalonia/issues/8123

**Comment by maxkatz6** [comment]
I would also recommend to try ICustomDrawOperation. Sample is here (Skia related, but you can use drawing context normally instead).
https://github.com/AvaloniaUI/Avalonia/blob/master/samples/RenderDemo/Pages/CustomSkiaPage.cs#L22

Main difference is that ICustomDrawOperation.Render is always executed on render thread, so you don't have access to the UI thread. I don't know if it will result in noticeably better perf, but you can try it out.

**Reply by JaggerJo** [reply]
Thx for the sample, switched to doing that. 👍

My problem seems to have shifted from 'Rendering is too slow' to 'Render is called but the screen not updated somehow'
